jupyter/d2l-java/chapter_multilayer-perceptrons/mlp-scratch.ipynb

```diff
@@ -0,0 +1,469 @@
+   "execution_count": 8,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "5"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "public NDArray net(NDArray X) {\n",
+    "\n",
+    "    X = X.reshape(new Shape(-1, 784));\n",
+    "    NDArray H = relu(X.dot(W1).add(b1));\n",
+    "    return H.dot(W2).add(b2);\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## The Loss Function\n",
+    "\n",
+    "To ensure numerical stability,\n",
+    "and because we already implemented\n",
+    "the softmax function from scratch\n",
+    "(:numref:`sec_softmax_scratch`),\n",
+    "we leverage Gluon's integrated function\n",
+    "for calculating the softmax and cross-entropy loss.\n",
+    "Recall our earlier discussion of these intricacies \n",
+    "(:numref:`sec_mlp`).\n",
+    "We encourage the interested reader \n",
+    "to examine the source code for `mxnet.gluon.loss.SoftmaxCrossEntropyLoss`\n",
+    "to deepen their knowledge of implementation details."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "6"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "Loss loss = Loss.softmaxCrossEntropyLoss();"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Training\n",
+    "\n",
+    "Fortunately, the training loop for MLPs\n",
+    "is exactly the same as for softmax regression.\n",
+    "\n",
+    "We run the training like how we did in Chapter 3, \n",
+    "(see :numref:`sec_softmax_scratch`),\n",
+    "setting the number of epochs to $10$ \n",
+    "and the learning rate to $0.5$."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "int numEpochs = 10;\n",
+    "float lr = 0.5f;\n",
+    "\n",
+    "float[] trainLoss;\n",
+    "float[] testAccuracy;\n",
+    "float[] epochCount;\n",
+    "float[] trainAccuracy;\n",
+    "\n",
+    "trainLoss = new float[numEpochs];\n",
+    "trainAccuracy = new float[numEpochs];\n",
+    "testAccuracy = new float[numEpochs];\n",
+    "epochCount = new float[numEpochs];"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "7"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "float epochLoss = 0f;\n",
+    "float accuracyVal = 0f;\n",
+    "\n",
+    "for (int epoch = 1; epoch <= numEpochs; epoch++) {\n",
+    "    \n",
+    "        System.out.print(\"Running epoch \" + epoch + \"...... \");\n",
+    "        // Iterate over dataset\n",
+    "        for (Batch batch : trainIter.getData(manager)) {\n",
+    "\n",
+    "            NDArray X = batch.getData().head();\n",
+    "            NDArray y = batch.getLabels().head();\n",
+    "\n",
+    "            for (NDArray param : params) {\n",
```

Comment:
Since parameters is not destroyed all the time, can you try to remove this attachment in every single loop?